File: epi_judge_java/epi/IsTreeSymmetric.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class IsTreeSymmetric {

    @EpiTest(testDataFile = "is_tree_symmetric.tsv")
    public static boolean isSymmetric(BinaryTreeNode<Integer> tree) {
        if (tree == null) return true;
        return isSymmetric(tree.left, tree.right);
    }

    private static boolean isSymmetric(BinaryTreeNode<Integer> left, BinaryTreeNode<Integer> right) {
        if (left == null && right == null) return true;
        if (left == null || right == null) return false;
        return left.data.equals(right.data)
                && isSymmetric(left.left, right.right)
                && isSymmetric(left.right, right.left);
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "IsTreeSymmetric.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/is_tree_symmetric.cc
#include "binary_tree_node.h"
#include "test_framework/generic_test.h"

bool eq(unique_ptr<BinaryTreeNode<int>> &node1, unique_ptr<BinaryTreeNode<int>> &node2);

bool IsSymmetric(const unique_ptr<BinaryTreeNode<int>> &tree) {
    if (tree == nullptr) return true;
    return eq(tree->left, tree->right);
}

bool eq(unique_ptr<BinaryTreeNode<int>> &node1, unique_ptr<BinaryTreeNode<int>> &node2) {
    if (node1 == nullptr || node2 == nullptr) return node1 == node2;

    if (node1->data != node2->data) return false;

    return eq(node1->left, node2->right) && eq(node1->right, node2->left);
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"tree"};
    return GenericTestMain(args, "is_tree_symmetric.cc", "is_tree_symmetric.tsv",
                           &IsSymmetric, DefaultComparator{}, param_names);
}
